target/linux/ramips/dts/mt7620n_hootoo_tm05.dts

```diff
@@ -0,0 +1,129 @@
+/dts-v1/;
+
+#include "mt7620n.dtsi"
+
+#include <dt-bindings/gpio/gpio.h>
+#include <dt-bindings/input/input.h>
+
+/ {
+	compatible = "hootoo,tm05", "ralink,mt7620n-soc";
+	model = "HooToo TM05";
+
+	chosen {
+		bootargs = "console=ttyS0,115200";
+	};
+
+	leds {
+		compatible = "gpio-leds";
+
+		green {
+			label = "tm05:green";
+			gpios = <&gpio2 0 GPIO_ACTIVE_HIGH>;
+			default-state = "on";
+			// Enabled below line, if flashing LED desired
+			//linux,default-trigger = "timer";
+		};
+
+		blue {
+			label = "tm05:blue";
+			gpios = <&gpio3 0 GPIO_ACTIVE_LOW>;
+			default-state = "on";
+		};
+	};
+
+	keys {
+		compatible = "gpio-keys";
+
+		reset {
+			label = "reset";
+			gpios = <&gpio2 1 GPIO_ACTIVE_LOW>;
+			linux,code = <KEY_RESTART>;
+		};
+	};
+};
+
+&gpio0 {
+	status = "okay";
+};
+
+&gpio1 {
+	status = "okay";
+};
+
+&gpio2 {
+	status = "okay";
+};
+
+&gpio3 {
+	status = "okay";
+};
+
+&i2c {
+	status = "okay";
+};
+
+&spi0 {
+	status = "okay";
+
+	m25p80@0 {
+		compatible = "jedec,spi-nor";
+		reg = <0>;
+		spi-max-frequency = <10000000>;
+
+		partitions {
+			compatible = "fixed-partitions";
+			#address-cells = <1>;
+			#size-cells = <1>;
+
+			partition@0 {
+				label = "u-boot";
+				reg = <0x0 0x30000>;
+				read-only;
+			};
+
+			partition@30000 {
+				label = "u-boot-env";
+				reg = <0x30000 0x10000>;
+				read-only;
+			};
+
+			factory: partition@40000 {
+				label = "factory";
+				reg = <0x40000 0x10000>;
+				read-only;
+			};
+
+			partition@50000 {
+				compatible = "denx,uimage";
+				label = "firmware";
+				reg = <0x50000 0x7b0000>;
+			};
+		};
+	};
+};
+
+&ehci {
+	status = "okay";
+};
+
+&ohci {
+	status = "okay";
+};
+
+&ethernet {
+	mtd-mac-address = <&factory 0x4>;
```

Comment:
What's at 0x28?

Reply:
Sorry, confused - why do you ask 0x28? I did pull the factory partition, confirm where the vendor programmed mac addresses are. Did this so the mac's match this way. OK?

Reply:
Typical assignment on mt7620 boards is
0x4 Wifi
0x28 LAN
0x2e WAN
(0x8004 2nd Wifi)

So, I would expect another MAC address in 0x28, which may be the same value as 0x4. So, is there an address at 0x28?

Reply:
Yep! Here is what I have (from factory partition),
```
0x0004: 00 1c c2 43 f6 b4  
0x0028: 00 1c c2 43 f6 b4  
0x002E: 00 1c c2 43 f6 b5  
0x8004: Nothing there, all 0xFF's
```
Any changes needed for this? Thanks!

Reply:
That's what I though. So, change &ethernet to 0x28 and put the following in your commit message:
```
MAC assignment based on vendor firmware:
2.4 GHz  *:b4   factory 0x4
LAN/label    *:b4   factory 0x28
WAN   *:b5   factory 0x2e
```

Maybe you can also align it more nicely than I did.

Reply:
Will do (alignment ... :-)). Put all three here, and the alias - then remove completely from 02_network? Just to make sure I get it right. Thanks!

Reply:
You cannot put wan MAC address into DTS, as it's assigned at runtime (eth0.2).

Reply:
OK, will do - thanks! Seems I only set LAN then (DTS, &ethernet) and WAN (02_network) ... not WiFi. Right?

Reply:
Wifi is set "automatically" by the calibration data. When using `ralink,mtd-eeprom`, the driver will expect the mac address at offset 4. With mtd-eeprom reading from 0x0, this will mean that the MAC address is read from 0x4 without having to specify it additionally.

For multiple-port devices, you then typically have LAN and WAN with different address, where LAN inherits from the eth0 address set in DTS and WAN is overwritten manually in 02_network. As you happen to have a one-port though (I overlooked this before), you don't need the WAN MAC address and can remove that part from 02_network indeed.

Have a look here for further reading: https://openwrt.org/docs/guide-developer/mac.address

How did you determine the WAN MAC address on vendor firmware in the first place, having only a one-port device?

Reply:
Perfect - that makes sense, thanks for explaining! To your question, I looked at the result from the stock firmware, tried to match that.

I think I have this correct in the code now - will update the commit, by all means comment if not. Just trying to get the Wi-Fi LED working, I think that's the only open one left (then will update the code for you to look at). Thanks!